PK: add `mbedtls_pk_write_pubkey_psa()`

valeriosetti opened the pull request:
## Description

Resolves #529

## PR checklist

- [x] **changelog** provided
- [x] **framework PR** not required
- [x] **mbedtls development PR** not required because: no change there
- [x] **mbedtls 3.6 PR** not required because: no backport
- **tests**  provided

valeriosetti committed:
drivers: pk: add mbedtls_pk_write_pubkey_psa()

Add function implementation, prototype and documentation.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: scripts: add tests for mbedtls_pk_write_pubkey_psa()

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
changelog: add note for mbedtls_pk_write_pubkey_psa()

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

davidhorstmann-arm commented on include/mbedtls/pk.h:
```diff
@@ -859,6 +859,39 @@ int mbedtls_pk_write_pubkey_pem(const mbedtls_pk_context *ctx, unsigned char *bu
+ *                  - #PSA_KEY_EXPORT_RSA_PUBLIC_KEY_MAX_SIZE(bitlen) for RSA keys,
+ *                  where the 'bitlen' parameter can be obtained through
+ *                  #mbedtls_pk_get_bitlen() on the same PK context.
+ * \param buf_len   Amount of bytes written into \p buf if the exporting
```
Nit: "Amount" -> "Number" is more accurate

mpg commented on include/mbedtls/pk.h:
```diff
@@ -859,6 +859,39 @@ int mbedtls_pk_write_pubkey_pem(const mbedtls_pk_context *ctx, unsigned char *bu
+ *
+ * \param ctx       PK context from which the public key is extracted. It must
+ *                  have been populated.
+ * \param buf       Output buffer where the pubblic key is written. It must not
```
Typo: pubBlic

mpg commented on include/mbedtls/pk.h:
```diff
@@ -859,6 +859,39 @@ int mbedtls_pk_write_pubkey_pem(const mbedtls_pk_context *ctx, unsigned char *bu
+ *                  operation is successful. In case of failure the value is 0.
+ *
+ * \return          0 if successful.
+ * \return          MBEDTLS_ERR_PK_BAD_INPUT_DATA if \p ctx, \p buf or \p buf_len
```
I don't think we need to check for that or promise that we do in the documentation. Generally speaking when passed NULL pointers that shouldn't be NULL, we're OK with getting whatever behaviour we get from the platform (like segfault on Linux etc).

Some of the old code in PK has more checks than our current policy, but I don't think we should copy that in new code.

(Though thanks for documenting every error condition here!)

valeriosetti replied:
While addressing comments I realized that the implementation I provided completely misses the support for opaque keys! This is not a regression compared to what there before since opaque contexts ([EC](https://github.com/Mbed-TLS/TF-PSA-Crypto/blob/cb9d0ed64831da3e7b85ea8741a57fdc27c010e6/drivers/builtin/src/pk_wrap.c#L712) and [RSA](https://github.com/Mbed-TLS/TF-PSA-Crypto/blob/cb9d0ed64831da3e7b85ea8741a57fdc27c010e6/drivers/builtin/src/pk_wrap.c#L785)) do not have the debug support. So the question: is this something we foresee that might be useful in the future? Or are we OK with the current implementation?
I think that extending the test coverage as [you suggested](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/577#discussion_r2584906619) will tell if at least we're fine with non-opaque keys, I guess

valeriosetti replied:
On a second thought I have a proposal/question as follow-up. How bad is it for opaque keys in PK to have their public part in the PK context (like non-opaque EC/RSA)? I cannot think of any drawback to this, but you know for sure better than me.

Rationale.
I found that `mbedtls_pk_ecc_set_pubkey_from_prv` and `mbedtls_pk_rsa_set_pubkey_from_prv` are basically the same, so instead of duplicating the code I can move them in `pk.c` with their prototype being in `pk_internal.h`. Then in `mbedtls_pk_write_pubkey_psa` I can check if `pk->pub_raw_len != 0`  and `memcpy` from that; however in case it's empty BUT `pk->priv_id` is not empty, I can try to export the public key from the private one (using the merge of `mbedtls_pk_ecc_set_pubkey_from_prv` & `mbedtls_pk_rsa_set_pubkey_from_prv` ) as last resort.
To me this makes `mbedtls_pk_write_pubkey_psa`  able to work in all scenarios without end user intervention and being able to support opaque keys as well!
Wdyt?

gilles-peskine-arm replied:
> How bad is it for opaque keys in PK to have their public part in the PK context (like non-opaque EC/RSA)?

It's just a memory-vs-performance compromise. Given that the current trend in pk is to reduce the number of cases when manipulating key representations, I think it would make sense to always store the public key in the context.

valeriosetti replied:
I just added a couple of commits ([[1]](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/577/commits/31c3c0c6c08de01437d4c974d5b0e734d6a4e19c) and [[2]](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/577/commits/a9f7d9e5e5a25e1968c2f08376520404803a720a)) to show what I had in mind.
I haven't extended testing though, so please wait a bit before reviewing the PR again.

mpg replied:
As Gilles said, and  as I was trying to say on the bitlen PR, I think we want to avoid as much as possible having fields in the `pk_context` that are sometimes set, sometimes not, and as maintainers we have to remember about them and what condition to check to know if the field is set or not, and that adds a lot of cognitive load when working with the code.

So, yes, every time we can reduce the number of cases we have to consider and move to a situation where the same information is always present in the same way not matter how the key was populated, I'll be in favour of doing so!

mpg commented on include/mbedtls/pk.h:
```diff
@@ -859,6 +859,39 @@ int mbedtls_pk_write_pubkey_pem(const mbedtls_pk_context *ctx, unsigned char *bu
+ *                  - #PSA_KEY_EXPORT_ECC_PUBLIC_KEY_MAX_SIZE(bitlen) for EC keys,
+ *                  - #PSA_KEY_EXPORT_RSA_PUBLIC_KEY_MAX_SIZE(bitlen) for RSA keys,
+ *                  where the 'bitlen' parameter can be obtained through
+ *                  #mbedtls_pk_get_bitlen() on the same PK context.
```
Thanks! This is an example of how we should always write documentation about buffer sizes :)

mpg commented on tests/suites/test_suite_pkwrite.function:
```diff
@@ -233,3 +233,66 @@ exit:
+    USE_PSA_INIT();
+    TEST_EQUAL(mbedtls_pk_parse_public_keyfile(&pk, key_file), 0);
+
+    /* TODO: Replace this with mbedtls_pk_get_type() once it will be available */
```
It has now been merged, so this can be updated :)

mpg commented on tests/suites/test_suite_pkwrite.function:
```diff
@@ -233,3 +233,66 @@ exit:
+    TEST_EQUAL(mbedtls_pk_write_pubkey_psa(&pk, buf, sizeof(buf), &buf_len),
+               MBEDTLS_ERR_PK_BAD_INPUT_DATA);
+
+    USE_PSA_INIT();
```
Post 1.0, we can (and should) just call `PSA_INIT()`. (Same with `PSA_DONE()`.)

mpg commented on tests/suites/test_suite_pkwrite.function:
```diff
@@ -233,3 +233,66 @@ exit:
+               MBEDTLS_ERR_PK_BAD_INPUT_DATA);
+
+    USE_PSA_INIT();
+    TEST_EQUAL(mbedtls_pk_parse_public_keyfile(&pk, key_file), 0);
```
Ok, that's one was to populate a context. As discussed elsewhere, we need to test other ways of populating the context. Also as discussed, that might either be done with more test functions, or my handling multiple methods in a single function.

Off the top of my head, I'd say:
- one function for parse public (this one)
- for parse keypair: possibly the same function, with an extra param for public/private, as I think the only thing that's going to change is the parse function being called
- possibly a single function from all 3 of `wrap_psa()` , `copy_from_psa()` and `copy_public_from_psa()`, as they all start with a PSA key?

mpg reviewed (changes requested):
Looks pretty good to me, thanks for the nice documentation, but I want more tests :)

valeriosetti committed:
drivers: pk: don't check for NULL pointers in mbedtls_pk_write_pubkey_psa()

It is OK to segfault (or any crash, depending on the platform) if a NULL
pointer is passed when the documentation clearly states that this must
not be done.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
drivers: pk: merge mbedtls_pk_(ecc|rsa)_set_pubkey_from_prv together

Since "mbedtls_pk_ecc_set_pubkey_from_prv" and
"mbedtls_pk_rsa_set_pubkey_from_prv" are exactly the same they can be
merged together.
The new function is named "mbedtls_pk_set_pubkey_from_prv"
and it's placed in "pk.c" because it's common to all key types. The
prototype was already in "pk_internal.h" so this was already fine.

Until now these functions has been used for non-opaque context, but given
the recent changes to PK, this new function can also be used to handle
opaque context in the same way (i.e. exporting its public key and keping
it in the PK context).

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
drivers: pk: enhance mbedtls_pk_write_pubkey_psa

Allow the function to be able to return the public key in all the scenarios
supported by PK:
- If the public key is already available in the PK context, then simply
  memcpy() it.
- If only the private key is available, then export the public key from
  the private one and then memcpy() it.

Advantages are:
- The user doesn't need to worry about filling the public part before
  calling this function.
- It allows the function to work also with opaque keys. This functionality
  wasn't supported before, so it's an improvement that comes for free.

The only side effect of this change is that the first parameter of
mbedtls_pk_write_pubkey_psa (i.e. mbedtls_pk_context) cannot be const
because otherwise it wouldn't be possible to pass it to
"mbedtls_pk_set_pubkey_from_prv" without some dirty casting.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti commented:
Sorry I had to rebase on `development` because I needed `mbedtls_pk_get_key_type()` in test code, but this has only been added recently in tf-psa-cryto.

valeriosetti committed:
tests: pkwrite: minor fixes to pk_write_pubkey_psa()

- replace USE_PSA_INIT/USE_PSA_DONE, with PSA_INIT/PSA_DONE
- use mbedtls_pk_get_key_type() to get the type of key being tested instead
  of the temporary solution that was added before when this function was
  not yet available.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
drivers: pk: pk_parse_key: RSA: export public key when parsing private key

For the EC keys this is already happening in "pk_parse_key_sec1_der()"
even though the procedure is a bit different in that case. For SEC1
the key might or might not be present on the parsed buffer, so it might
or might not be derived form the private counterpart. For RSA keys there
is no such possibility: the public key should always be exported from the
private key.

The point here is that at the end of the "mbedtls_pk_parse_key()" we want
private and public fields of the PK context to be filled for all the
supported key types.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: pkwrite: extend pk_write_pubkey_psa() to test also private keys

In this way both private and public key files are parsed and tested with
the same test function.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: pkwrite: rename pk_write_pubkey_psa() test function

Since we're going to add new test scenarios for
"mbedtls_pk_write_pubkey_psa()" where the PK context is generated in a
different way than parsing a key file, we rename "pk_write_pubkey_psa"
to "pk_write_pubkey_psa_from_file" to differentiate it from the new test
functions that will be added in following commits.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: pkwrite: move testing part of pk_write_pubkey_psa() out of pk_write_pubkey_psa_from_file()

This will become handy in follow up commits where the same core test
function will be used elsewhere.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: pk: add pk_write_pubkey_psa_from_psa()

This is meant to extend testing of "mbedtls_pk_write_pubkey_psa()" by
creating a PK context starting from a PSA key with:

- pk_wrap_psa
- pk_copy_from_psa
- pk_copy_public_from_psa

As for the already existing test cases, 2 EC and 2 RSA keys are tested.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: pkwrite: minor typo adjustments to test case labels

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
framework: update reference

This is not needed at all by this PR, but it required to make the CI
happy. It's just updating the reference of framework to the current HEAD
of main branch.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

mpg commented on drivers/builtin/src/pk.c:
```diff
@@ -143,6 +143,23 @@ int mbedtls_pk_setup(mbedtls_pk_context *ctx, const mbedtls_pk_info_t *info)
+
+    /* Public key already available in the PK context. Nothing to do. */
+    if (pk->pub_raw_len > 0) {
+        return PSA_SUCCESS;
```
Minor: since the function returns `int`, I'd find `return 0` more natural here.

mpg commented on drivers/builtin/src/pkwrite.c:
```diff
@@ -478,13 +478,24 @@ int mbedtls_pk_write_key_der(const mbedtls_pk_context *key, unsigned char *buf,
  * Public functions for public keys in "PSA friendly" format.
  ******************************************************************************/
-int mbedtls_pk_write_pubkey_psa(const mbedtls_pk_context *ctx, unsigned char *buf,
+int mbedtls_pk_write_pubkey_psa(mbedtls_pk_context *ctx, unsigned char *buf,
```
This feels a bit hacky: having to remove the `const` is a bad smell. I'd rather make sure all functions that populate a context always set the public key field (presumably by calling `mbedtls_pk_set_pubkey_from_prv()`.

This goes with my general wish that once a context is populated, all the fields from `pk_context` are set and usable (except for `priv_id` when the context is only holding public key material).

Unless you think there's a reason this can't be done?

valeriosetti replied:
> This feels a bit hacky: having to remove the const is a bad smell

TBH I was skeptical as well about this change. I checked current calls to `mbedtls_pk_debug` in TLS code and in all the cases a `const` pointer is being passed. It makes sense of course, because we're only supposed to get some data out of the PK context, not to modify it.
Also I knew that this change was "requesting" a bit too many weird castings in the TLS code and that this was not nice as well.

> I'd rather make sure all functions that populate a context always set the public key field (presumably by calling mbedtls_pk_set_pubkey_from_prv().

I'll look into improving this.

mpg commented on tests/suites/test_suite_pkwrite.data:
```diff
@@ -154,18 +154,34 @@ Derive public key EC X448
+Export public key PSA - Parse public EC key secp256r1
 depends_on:PSA_WANT_KEY_TYPE_ECC_PUBLIC_KEY:PSA_WANT_ECC_SECP_R1_256
-pk_write_pubkey_psa:"../framework/data_files/ec_256_pub.der"
+pk_write_pubkey_psa:"../framework/data_files/ec_256_pub.der":0
```
Minor: can we have symbolic names, like `MBEDTLS_PK_TEST_{PUBLIC,PRIVATE}` (either using `#define`s or an enum)? We already have too many test with 0/1 parameters, and as long as there's only one we can often guess, but when there's more than one it's uncomfortable to read.

In the spirit that you suggested of raising the bar for readability in newly added tests :)

valeriosetti replied:
Sure, that makes completely sense. I usually hate when I have to switch back and forth between `.data` and `.function` files because there are sequences of `1` and `0` in the test data and my bad short term memory cannot remember the order of parameters in test functions :)

gilles-peskine-arm replied:
> I usually hate when I have to switch back and forth between .data and .function files because there are sequences of 1 and 0 in the test data and my bad short term memory cannot remember the order of parameters in test functions

That calls for IDE support! Emacs users can use [mbedtls-test-data-mode](https://github.com/Mbed-TLS/mbedtls-docs/blob/main/tools/emacs/mbedtls-test-data-mode.el). Contributions for other editors are welcome.

mpg replied:
I thought @davidhorstmann-arm has something from vim using a small special-purpose LSP, but I can't seem to find it again.

Still, I think we can all agree that symbolic names have value regardless of the level of editor/IDE support :)

mpg commented on tests/suites/test_suite_pkwrite.function:
```diff
@@ -295,3 +338,36 @@ exit:
+        default:
+            TEST_FAIL("Unknown method");
+    }
+
```
Shouldn't we call `test_pk_write_pubkey_psa()`? Am I missing something obvious or did you?

valeriosetti replied:
Oops, bad mistake, sorry

mpg commented on tests/suites/test_suite_pkwrite.function:
```diff
@@ -295,3 +338,36 @@ exit:
+
+    key_id = make_psa_key_from_predefined(key_type, key_bits);
+
+    switch (pk_gen_method) {
```
Note: this is likely to be also useful for other test functions, so it would make sense turning that into its own helper function. (Can be done either in this PR, or in the next PR that needs something similar - I think perhaps #575 could use something similar.)

valeriosetti replied:
Since we also collaborated on previous PRs about PK I guess you saw that this was a duplication of the test helper function that's already used in [test_suite_pk.function](https://github.com/Mbed-TLS/TF-PSA-Crypto/blob/348d0eeb40bfc053225d4f1da69b60644c1d3110/tests/suites/test_suite_pk.function#L179), so I agree that creating a common helper makes sense.

valeriosetti replied:
The only thing I'm not sure about is if it's better to do this in this PR or in a quick follow-up one. This one is going to be in the range of 20 commits already, so perhaps a quick follow-up one is better.

valeriosetti replied:
https://github.com/Mbed-TLS/TF-PSA-Crypto/issues/598

mpg reviewed (changes requested):
Thanks for adding the bunch of tests I requested, I like how your structured them too :) 

However in addition to small suggestions I have two blocker comments:
- Possibly mutating the context seems dirty: please either remove this, or convince me it's the right way to go (but I'll be hard to convince, fair warning).
- Apparently missing actual testing for contexts creating from PSA - or am I just not seeing it?

Also, a few other comments/suggestions that are optional.

mpg commented on tests/suites/test_suite_pkwrite.data:
```diff
@@ -153,3 +153,83 @@ pk_write_public_from_private:"../framework/data_files/ec_x25519_prv.der":"../fra
+
+Export public key PSA - Parse public - EC key secp256r1
+depends_on:PSA_WANT_KEY_TYPE_ECC_PUBLIC_KEY:PSA_WANT_ECC_SECP_R1_256
+pk_write_pubkey_psa_from_file:"../framework/data_files/ec_256_pub.der":0
```
I thought I left a comment here but it seems github lost it: I'd like a symbolic name instead of 0/1 for the new argument, like you did for `pk_write_pubkey_psa_from_psa()`.

valeriosetti committed:
drivers: pk: fix return code in mbedtls_pk_set_pubkey_from_prv()

PK returns "int" not "psa_status_t" so the return code should have been
"0" and not "PSA_SUCCESS".

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
drivers: pkwrite: do not modify PK context in mbedtls_pk_write_pubkey_psa

Partially reverts commit 63660b73a4cd0c3e02a0edc3c7bacb913dfaee34 for
what concerns with the following:

- keep "const" on the PK parameter being passed in input to the function;
- do not try to complete the PK context if the public key is missing.

The rationale here is that this is just a debug function meant to get some
data from the PK context. It's the caller responsibility to properly fill
the PK context before calling it.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: pkwrite: add missing test part in pk_write_pubkey_psa_from_psa()

After the PK context has been initialized it must be tested (obviously),
the same as in "pk_write_pubkey_psa_from_file", but this part was missing.
This commit fixes this.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: pkwrite: fix checks in test_pk_write_pubkey_psa

Add a previously missing check that caused some tests to erroneously fail
when "pk->pub_raw_len = 0" and also the current output buffer len was
set to 0.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
drivers: pk: set public key in PK context in mbedtls_pk_wrap_psa()

Add a call to "mbedtls_pk_set_pubkey_from_prv" from "mbedtls_pk_wrap_psa"
in order to set the public key in PK context when the PSA key pair is
wrapped.

This aligns "mbedtls_pk_wrap_psa" to all other current ways of generating
a PK context from a key pair: "mbedtls_pk_copy_from_psa" and
"mbedtls_pk_parse_keyfile".

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: pkwrite: add enum to specify the type of key file being parsed

This helps reading the ".data" file so that it's immediately clear if
a certain file being parsed is supposed to contain a private or public
key.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

mpg reviewed (changes requested):
Thanks for addressing my comments. Looks almost all good to me, but I think we're missing a test for calling the new function on an unpopulated context: 
```c
unsigned char buf[PSA_EXPORT_PUBLIC_KEY_MAX_SIZE];
mbedtls_pk_init(&ctx);
TEST_EQUAL(MBEDTLS_ERR_PK_BAD_INPUT_DATA, mbedtls_pk_write_pubkey_psa(&ctx, buf, sizeof(buf), &buflen));
```
(or similar, without the silly mistake I probably made).

I'd suggest adding a small dedicated test function for that. That way, looking at the `.data` file it's very apparent that we're testing all the cases for this function: unpopulated context (negative), then all 5 methods of populating the context (4 positive cases each).

valeriosetti committed:
tests: pk: test mbedtls_pk_write_pubkey_psa() on unpopulated PK context

This completes the testing of mbedtls_pk_write_pubkey_psa() for all
possible scenarios.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

mpg reviewed (dismissed):
I'm finally satisfied with the amount of testing ;)

valeriosetti commented:
@mpg @bjwtaylor I had to rebase to resolve a conflict on the `framework`. Can you please re-check?

mpg reviewed (approved):
LGTM, thanks!
